### PR TITLE
Added example views.py for business logic template server

### PR DESCRIPTION
**Pull request by matthewnickolai**
Added an example of adding/modifying slots using business logic and
business logic transitions in /my_project/my_app/views.py .

### CODE REVIEW

**Comment by ypkang** [comment]
@matthewnickolai  Could you tweak the MR title? Also, I don't think there's a need for using the [][] format/prefix on here.

File: Dockerfile [hunk 1]
```diff
@@ -1,6 +1,10 @@
 FROM python:2.7
 ENV PYTHONUNBUFFERED 1
 
+# build arguments
+ARG LUCIDA_HOSTNAME
+ENV LUCIDA_HOSTNAME $LUCIDA_HOSTNAME
```
Comment (ypkang) [inline Dockerfile]: Why do you need these?

File: README.md [hunk 2]
```diff
@@ -13,7 +13,308 @@ docker-compose up
 ```
 
 # Business logic interface
-TBA
+
+## Business Logic Setup
```
Comment (ypkang) [inline README.md]: We don't need this section for this repo right?

File: my_project/my_app/hotels.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+def find_express_deal (location, price):
+
+    print ('@' * 40)
+    print (location)
+    print (price)
+    print ('@' * 40)
```
Comment (ypkang) [inline my_project/my_app/hotels.py]: Remove debug print statements.

**Comment by matthewnickolai** [comment]
@ypkang better?

File: my_project/my_app/views.py [hunk 4]
```diff
@@ -1,23 +1,88 @@
...
+                            }
+                        ],
+                        "required_matches": "EQ 1",
+                        "type": "string"
+                    }
+                    request.data['slots']['_HOTEL_TYPE_'] = _HOTEL_TYPE_
+
+        # print out the modified business logic payload
+        print "Running through the webhook"
+        print json.dumps(request.data, indent=2)
```
Comment (ypkang) [inline my_project/my_app/views.py]: Remove print statement.

File: my_project/my_app/views2.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
+"""
```
Comment (ypkang) [inline my_project/my_app/views2.py]: What's this file?

**Review by ypkang — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 1 above)
Comment (matthewnickolai) [reply Dockerfile]: I likely don't, I just copied these from the original business logic template server

File: README.md (hunk 2 above)
Comment (matthewnickolai) [reply README.md]: I suppose we don't technically /need/ it, but I don't think it will hurt.  Are we planning on giving this to people to use /without/ spotlight?

File: my_project/my_app/hotels.py (hunk 3 above)
Comment (matthewnickolai) [reply my_project/my_app/hotels.py]: removed

File: my_project/my_app/views.py (hunk 4 above)
Comment (matthewnickolai) [reply my_project/my_app/views.py]: I actually think printing the json.dumps of the payload could have value in an introductory template server.  I can remove them if you'd rather, though.

File: my_project/my_app/views2.py (hunk 5 above)
Comment (matthewnickolai) [reply my_project/my_app/views2.py]: This file is an example of business logic transitions.  It is separate than the `views.py`, which just shows off adding/modifying slots.

File: README.md (hunk 2 above)
Comment (ypkang) [reply README.md]: I don't think it makes sense to to ship this server to people without spotlight. Let's remove this section to avoid confusion.

File: my_project/my_app/views2.py (hunk 5 above)
Comment (ypkang) [reply my_project/my_app/views2.py]: I'd merge this into the the other one.

**Review by ypkang — CHANGES_REQUESTED** [review]

File: my_project/my_app/views.py [hunk 6]
```diff
@@ -1,23 +1,86 @@
...
-
-        Accepts a post request with two integers
-        and returns a reponse containing their sum.
         """
-        int1, int2 = (int(value) for value in request.data.values())
-        result = int1 + int2
-        return Response(result)
+        #overwrite state for state transition example
+        if request.data['state'] == "clean_hello":
+            request.data['state'] = "clean_goodbye"
```
Comment (ypkang) [inline my_project/my_app/views.py]: Could you design a slightly better example for this? Instead of just override the state if the current state is clean_hello, have a condition related to slots and states combination.

File: docker-compose.yml [hunk 7]
```diff
@@ -2,7 +2,8 @@ version: '2'
 
 services:
   web:
-    build: .
+#    build: .
+    image: business-logic
```
Comment (ypkang) [inline docker-compose.yml]: Why is this added?

**Review by ypkang — CHANGES_REQUESTED** [review]

File: docker-compose.yml (hunk 7 above)
Comment (matthewnickolai) [reply docker-compose.yml]: It was just copied from the original business logic server template.  Reverted.

File: my_project/my_app/views.py (hunk 6 above)
Comment (matthewnickolai) [reply my_project/my_app/views.py]: I changed this to something more arbitrary and made up.  I think we should maybe remove 'hotels.py' if we are planning on giving them something they can't immediately test in Spotlight

File: my_project/my_app/views.py [hunk 8]
```diff
@@ -1,23 +1,92 @@
...
+                for candidate in range(len(slot_data['candidates'])):
+                    request.data['slots'][slot]['candidates'][candidate]['resolved'] = 1
+                    if slot <> '_DATE_':
+                        request.data['slots'][slot]['candidates'][candidate]['value'] = \
+                            request.data['slots'][slot]['candidates'][candidate]['tokens']
+            else:
+                request.data['slots'][slot]['resolved'] = 1
+
+        #magical API call to check their credit
+        available_credit = check_available_credit()
```
Comment (ypkang) [inline my_project/my_app/views.py:82]: I think we should implement a very simple version of this function (e.g. literally just return a static number). I see your point that this example won't be useful without the corresponding competencies created in the Platform but the example on its own should be functioning given the right input. If someone just stand up this server and querying it with curl, it should work.

File: my_project/my_app/views.py [hunk 9]
```diff
@@ -1,23 +1,92 @@
...
+
+        # this loop sets all of the _SLOTS_ to have a `"resovled": 1` so they will be kept
+        # through each turn of the conversation.  Currently, each turn the slots are sent
+        # with a `"resolved": -1`, so they need to be reset each time, however, they are
+        # changing to be persistent based on their resolved status in an update coming soon
+        for (slot, slot_data) in request.data['slots'].iteritems():
+            if 'candidates' in request.data['slots'][slot]:
+                for candidate in range(len(slot_data['candidates'])):
+                    request.data['slots'][slot]['candidates'][candidate]['resolved'] = 1
+                    if slot <> '_DATE_':
```
Comment (ypkang) [inline my_project/my_app/views.py]: just `!=` instead of `<>`?

File: my_project/my_app/views.py [hunk 10]
```diff
@@ -1,23 +1,92 @@
...
 from rest_framework.views import APIView
 from rest_framework.response import Response
 from rest_framework import status  # pylint: disable=unused-import
 
 
-class Sum(APIView):
+class BusinessLogic(APIView):
     """
-    Sample view class.
+    This is an example of adding/modifying slots using business logic, using
```
Comment (ypkang) [inline my_project/my_app/views.py:13]: I suggest briefly describing the competency design. For example, is it a informational/confirmational, which slots does it have, what are the common conversational scenarios. I think it helps people understand the example better.

File: my_project/my_app/views.py [hunk 11]
```diff
@@ -1,23 +1,92 @@
...
+            request.data['state'] = 'credit_card_offer'
+            return Response(request.data)
+
+        if transaction_type == 'express deal':
+            if location and price:
+                # This is our magical API call to find express deals
+                hotel = find_express_deal(location, price)
+                if hotel:
+                    # This is how to add new _SLOTS_ to the business logic json
+                    _HOTEL_RATING_ = {
```
Comment (ypkang) [inline my_project/my_app/views.py]: I know you are trying to mimic the slot naming convention but this is a pretty weird variable name. I'd suggest either `HOTEL_RATING` or `hotel_rating`.

File: my_project/my_app/views.py [hunk 12]
```diff
@@ -1,23 +1,92 @@
...
+                            {
+                                "resolved": 1,
+                                "value": hotel['hotel_rating']
+                            }
+                        ],
+                        "required_matches": "EQ 1",
+                        "type": "string"
+                    }
+                    request.data['slots']['_HOTEL_RATING_'] = _HOTEL_RATING_
+                    _HOTEL_TYPE_ = {
```
Comment (ypkang) [inline my_project/my_app/views.py]: Same here.

**Review by ypkang — CHANGES_REQUESTED** [review]

File: my_project/my_app/views.py (hunk 8 above)
Comment (matthewnickolai) [reply my_project/my_app/views.py:82]: I did add one to `hotels.py` already

Comment (ypkang) [reply my_project/my_app/views.py:82]: Oh just noticed that you do have an implementation for this.

**Review by ypkang — APPROVED** [review]
